File: puremvc/puremvc/MVC/Patterns/Command/MacroCommand.cs
using System;
using System.Collections.Generic;
using MVC.Interfaces;
using MVC.Patterns.Observer;

namespace MVC.Patterns.Command
{
    public class MacroCommand : Notifier, ICommand
    {
        public MacroCommand()
        {
            subcommands = new List<Func<ICommand>>();
            InitializeMacroCommand();
        }
        protected virtual void InitializeMacroCommand()
        {
        }
        protected void AddSubCommand(Func<ICommand> factory)
        {
            subcommands.Add(factory);
        }
        public virtual void Execute(INotification notification)
        {
            while(subcommands.Count > 0){
                var factory = subcommands[0];
                commandInstance.InitializeNotifier(MultitonKey);
                commandInstance.Execute(notification);
                subcommands.RemoveAt(0);
            }
        }
        public readonly IList<Func<ICommand>> subcommands;
    }
}

File: framework/project/base/XScripts/Frameworks/MVC/Patterns/Command/MacroCommand.lua
local MacroCommand = Class("MVC.Command.MacroCommand", MVC.Notifier)
function MacroCommand:Ctor()
    self.super:Ctor(self)
    self.subCommands = {}
    self:InitializeMacroCommand()
end

function MacroCommand:InitializeMacroCommand()
end

function MacroCommand:AddSubCommand(commandClassRef)
    table.insert(self.subCommands, commandClassRef)
end

function MacroCommand:Execute(note)
    while (#self.subCommands > 0) do
        local ref = table.remove(self.subCommands, 1)
        local cmd = ref.new()
        cmd:InitializeNotifier(self.multitonKey)
        cmd:Execute()
    end
end

return MacroCommand